src/Compilers/Core/Portable/Compilation/IExpression.cs

```diff
@@ -0,0 +1,776 @@
+﻿using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.Semantics
+{
+    /// <summary>
+    /// Represents a C# or VB expression.
+    /// </summary>
+    public interface IExpression : IOperation
+    {
+        /// <summary>
+        /// Result type of the expression.
+        /// </summary>
+        ITypeSymbol ResultType { get; }
+        /// <summary>
+        /// If the expression evaluates to a constant value, the value of the expression, and otherwise null.
+        /// </summary>
+        object ConstantValue { get; }
+    }
+
+    /// <summary>
+    /// Represents a C# or VB method invocation.
+    /// </summary>
+    public interface IInvocation : IExpression
+    {
+        /// <summary>
+        /// Method to be invoked.
+        /// </summary>
+        IMethodSymbol TargetMethod { get; }
+        /// <summary>
+        /// 'This' or 'Me' argument to be supplied to the method.
+        /// </summary>
+        IExpression Instance { get; }
+        /// <summary>
+        /// Kind of the invocation.
+        /// </summary>
+        InvocationKind InvocationKind { get; }
+        /// <summary>
+        /// Arguments of the invocation, excluding the instance argument.
+        /// </summary>
+        ImmutableArray<IArgument> Arguments { get; }
+        /// <summary>
+        /// Find the argument supplied for a given parameter of the target method.
+        /// </summary>
+        /// <param name="parameter">Parameter of the target method.</param>
+        /// <returns>Argument corresponding to the parameter.</returns>
+        IArgument ArgumentMatchingParameter(IParameterSymbol parameter);
+    }
+
+    /// <summary>
+    /// Kinds of invocations.
+    /// </summary>
+    public enum InvocationKind
+    {
+        /// <summary>
+        /// Virtual invocation.
+        /// </summary>
+        Virtual,
+        /// <summary>
+        /// Static or shared invocation with no instance value.
+        /// </summary>
+        Static,
+        /// <summary>
+        /// Non-virtual invocation with an instance value.
+        /// </summary>
+        NonVirtualInstance
+    }
+
+    /// <summary>
+    /// Represents an argument in a method invocation.
+    /// </summary>
+    public interface IArgument
```

Comment:
Kind is argument metadata, because it refers to how the arguments are specified in the call. An argument matching a particular parameter might be specified positionally in one call and by name in another. One might argue that this is not important for semantics and so mode should be omitted.

Mode is parameter metadata and should be replaced with a reference to the corresponding parameter. This would also allow omitting names from specified-by-name arguments.